streams/src/main/java/org/apache/kafka/streams/processor/internals/AssignedTasks.java

```diff
@@ -0,0 +1,459 @@
+            } else {
+                log.trace("{} couldn't resume task {} assigned partitions {}, task partitions", logPrefix, taskId, partitions, task.partitions());
+            }
+        }
+        return false;
+    }
+
+    private void transitionToRunning(final Task task) {
+        running.put(task.id(), task);
+        for (TopicPartition topicPartition : task.partitions()) {
+            runningByPartition.put(topicPartition, task);
+        }
+        for (TopicPartition topicPartition : task.changelogPartitions()) {
+            runningByPartition.put(topicPartition, task);
+        }
+    }
+
+    Task runningTaskFor(final TopicPartition partition) {
+        return runningByPartition.get(partition);
+    }
+
+    Set<TaskId> runningTaskIds() {
+        return running.keySet();
+    }
+
+    Map<TaskId, Task> runningTaskMap() {
+        return Collections.unmodifiableMap(running);
+    }
+
+    public String toString(final String indent) {
+        final StringBuilder builder = new StringBuilder();
+        describe(builder, running.values(), indent, "Running:");
+        describe(builder, suspended.values(), indent, "Suspended:");
+        describe(builder, restoring.values(), indent, "Restoring:");
+        describe(builder, created.values(), indent, "New:");
+        return builder.toString();
+    }
+
+    private void describe(final StringBuilder builder,
+                          final Collection<Task> tasks,
+                          final String indent,
+                          final String name) {
+        builder.append(indent).append(name);
+        for (final Task t : tasks) {
+            builder.append(indent).append(t.toString(indent + "\t\t"));
+        }
+        builder.append("\n");
+    }
+
+    private List<Task> allInitializedTasks() {
+        final List<Task> tasks = new ArrayList<>();
+        tasks.addAll(running.values());
+        tasks.addAll(suspended.values());
+        tasks.addAll(restoring.values());
+        return tasks;
+    }
+
+    Collection<Task> restoringTasks() {
+        return Collections.unmodifiableCollection(restoring.values());
+    }
+
+    Set<TaskId> allAssignedTaskIds() {
+        final Set<TaskId> taskIds = new HashSet<>();
+        taskIds.addAll(running.keySet());
+        taskIds.addAll(suspended.keySet());
+        taskIds.addAll(restoring.keySet());
+        taskIds.addAll(created.keySet());
+        return taskIds;
+    }
+
+    void clear() {
+        runningByPartition.clear();
+        running.clear();
+        created.clear();
+        suspended.clear();
+        restoredPartitions.clear();
+    }
+
+    Set<TaskId> previousTaskIds() {
+        return previousActiveTasks;
+    }
+
+    void commit() {
+        performTimedAction(commitAction);
+    }
+
+    void maybeCommit() {
+        performTimedAction(maybeCommitAction);
+    }
+
+    private void performTimedAction(final TaskAction action) {
+        timerStartedMs = time.milliseconds();
```

Comment:
One meta comment: right now the latency recording logic is scattered in both `AssignedTasks` and `StreamThread`. While we cannot move all recording logic out of `StreamThread` because it needs to time the `poll` call. One idea I had is the following:

Still maintain all the recordings in the stream thread since it is for the thread-level metrics anyways. To do that we need to do the following:

0. We only time the total time for each call iterating over all the tasks in `StreamThread`.
1. For `process()` which includes the stream-time based `punctuate`, we extract the punctuate logic out of it (see below) and get the number of records processed, so we can record the latency as `time / num_processed`.
2. For `commit` we know each task will commit, so we can record the latency as `time / num_tasks`.
3. For `maybeCommit` we return a integer of the number of tasks that actually call commits, and record as `time / num_commit_calls`.
4. For both stream/system-time based `punctuate`, we will do that after the `process()` call within the `do-while` loop in StreamThread, that is, after one iteration of tasks processing, we check if punctuate can be triggered based on both times. Since each task only processed at most once after `process()`, stream-time punctuation correctness is preserved, and since we check system-time based punctuation more often it would be more accurate on timing as well. In addition, we let this call to return how many `punctuate` calls are actually triggered (note some task may call twice within one loop, one for stream time and one for system time), and record latency as `time / num_total_punc_calls`.

So we still keep the metrics recording in StreamThread, plus the latency calculations now becomes a bit coarsen grained, which sounds a good trade-off for performance. WDYT?